elastic/apm-server
# decoder: Always map transaction.id

## Motivation/summary

Removes the need for the `transaction` object to be set in the intake
payload in order to map the `transaction_id` to `transaction.id`. Any
incoming events with `transaction_id` set, but `transaction` unset will
now correctly have `transaction.id` set.

## Checklist

~- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)~
~- [ ] Documentation has been updated~

## How to test these changes

1. `docker-compose up -d`.
2. `cd systemtest && go run cmd/runapm/main.go`
3. `curl -H 'Content-type: application/x-ndjson' --data-binary @testdata/intake-v2/errors_transaction_id.ndjson "http://localhost:60159/intake/v2/events"` Replace port `60159` with the actual port from `runapm`.
4. Issue the following query and ensure it matches the document:
```
GET /logs-apm.error-default/_search
{
  "fields": [
    "transaction.id"
  ], 
  "_source": false, 
  "query": {
    "match": {
      "transaction.id": "1234567890987654"
    }
  }
}
```

```javascript
{
  "took": 9,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "max_score": 0.2876821,
    "hits": [
      {
        "_index": ".ds-logs-apm.error-default-2022.08.08-000001",
        "_id": "AAKlfIIByzLEhspeH8SC",
        "_score": 0.2876821,
        "fields": {
          "transaction.id": [
            "1234567890987654"
          ]
        }
      }
    ]
  }
}
```

## Related issues

Closes #8664
